Functions/Mock.ps1

```diff
@@ -1659,7 +1659,15 @@ function New-BlockWithoutParameterAliases {
     try {
         if ($PSVersionTable.PSVersion.Major -ge 3) {
             $params = $Metadata.Parameters.Values
-            $ast = $Block.Ast.EndBlock
+            if ($Block.Ast -is [System.Management.Automation.Language.FunctionDefinitionAst]) {
```

Comment:
Could this be moved to a custom validation attribute?

If so, we could then apply the attr to both the `Mock` and `Assert-MockCalled` functions to catch problems even sooner

Reply:
To be sure we are talking about the same thing, you want a custom validator to set against to set against the $ScriptBlock Param, like this:

    Param(
      [ValidateNotNullOrEmpty()]
      [SetScriptBlockToAST()]
      $ScriptBlock
    )

If I'm not mistaken - for the custom validator, you need to either define this in c# or write a Powershell class. If Pester is still trying to support older version of Powershell - this will have to be c# - unless there is some other way to create a validator or I misunderstood what you meant.

If you want to be able to use this logic elsewhere - I personally would lean towards justing make this logic into it's own function (maybe Get-ScriptBlockAst or similar?). Then, this case would look like:

    Param( [ScriptBlock]$ScriptBlock )
    # blah blah
    $ast  = Get-ScriptBlockAst $ScriptBlock
    $blockText = $ast.Extent.Text

Reply:
Good point. I tend to get carried away with the features of PS 5+ :).

Actually, I meant moving the if checks only, so they could be use to validate the parameter value. We would still need to do the check here inside the function to properly extract **EndBlock** property if it a `FunctionDefinitionAst` or a `ScriptBlockAst`.

On the other hand future changes could make the validation function and the if statements here get out of sync, so I think we should hold off for now.

Reply:
@renehernandez Are you sure? I'm thinking of just a trivial function, where we receive a scriptblock, and then always return a [ScriptBlockAst] . Then, you don't have to verify the type in any of the Mock functions.

    # pseudo
    Function Get-ScriptBlockAst {
      param ( [scriptblock]$Scriptblock )
      
      switch ($scriptblock) {
        $_.ast -is [FunctionDefinitionAst] { $returnblock = $_.body }
        $_.ast -is [ScriptBlockAst] { $returnblock = $_ }
        $_.ast -is [SomeOtherAst] { $returnblock = $_.some.path.body }     
        Default { Throw = "You have a magical block, who defies logic. Cherish him!" }

     Return $returnblock
    } # End Func

@nohwnd @renehernandez Thoughts on this? If you want to do it this way, I can modify the PR to match this.

Reply:
The conversion function seems like a nice way to do it.

Reply:
@clcaldwell The new function would help simplify the logic in `New-BlockWithoutParameterAliases` function, which I am all for it, but we would not be able to reuse it within the `ValidationScript` block (the original purpose of my comment) since the function does not return boolean.

Reply:
Ah, I see what you are saying now. I'll modify the PR to have this as a function just so this is a bit cleaner - and then let's get this merged if we can, the bug is squashed. 

I'll open a separate ticket to have a discussion on potential refactoring for scriptblock validation/conversion across the board in Pester.

Reply:
@renehernandez validation script errors suck, it's much nicer to have a guard at the start of the function that returns a more reasonable message instead :)